src/main/java/rs/teslaris/core/service/impl/DocumentFileServiceImpl.java

```diff
@@ -0,0 +1,205 @@
+        documentFileRepository.delete(documentToDelete);
+    }
+
+    private String detectMimeType(MultipartFile file) {
+        var contentAnalyzer = new Tika();
+
+        String trueMimeType;
+        String specifiedMimeType;
+        try {
+            trueMimeType = contentAnalyzer.detect(file.getBytes());
+            specifiedMimeType =
+                Files.probeContentType(Path.of(Objects.requireNonNull(file.getOriginalFilename())));
+        } catch (IOException e) {
+            throw new StorageException("Failed to detect mime type for file.");
+        }
+
+        if (!trueMimeType.equals(specifiedMimeType) &&
+            !(trueMimeType.contains("zip") && specifiedMimeType.contains("zip"))) {
+            throw new StorageException("True mime type is different from specified one, aborting.");
+        }
+
+        return trueMimeType;
+    }
+
+    @Override
+    public void parseAndIndexPdfDocument(DocumentFile documentFile, MultipartFile multipartPdfFile,
+                                         String serverFilename, DocumentFileIndex documentIndex) {
+        if (!isPdfFile(multipartPdfFile)) {
+            return;
+        }
+
+        var documentContent = extractDocumentContent(multipartPdfFile);
+        var documentTitle = extractDocumentTitle(multipartPdfFile);
+
+        var contentLanguageDetected = detectLanguage(documentContent);
+        var titleLanguageDetected = detectLanguage(documentTitle);
+
+        saveDocumentIndex(documentContent, documentTitle, contentLanguageDetected,
+            titleLanguageDetected, documentFile, serverFilename, documentIndex);
+
+        documentFileIndexRepository.save(documentIndex);
+    }
+
+    private boolean isPdfFile(MultipartFile multipartFile) {
+        return Objects.equals(multipartFile.getContentType(), "application/pdf");
+    }
+
+    private String extractDocumentContent(MultipartFile multipartPdfFile) {
+        String documentContent;
+        try (var pdfFile = multipartPdfFile.getInputStream()) {
+            var pdDocument = PDDocument.load(pdfFile);
+            var textStripper = new PDFTextStripper();
+            documentContent = textStripper.getText(pdDocument);
+            pdDocument.close();
+        } catch (IOException e) {
+            throw new LoadingException("Error while trying to load PDF file content.");
+        }
+        return documentContent;
+    }
+
+    private String extractDocumentTitle(MultipartFile multipartPdfFile) {
+        var originalFilename = Objects.requireNonNull(multipartPdfFile.getOriginalFilename());
+        return originalFilename.split("\\.")[0];
+    }
+
+    private String detectLanguage(String text) {
+        LanguageDetector languageDetector;
+        try {
+            languageDetector = LanguageDetector.getDefaultLanguageDetector().loadModels();
+        } catch (IOException e) {
+            throw new NotFoundException("Error while loading language models.");
+        }
```

Comment:
Can this make a performance issue? Can we loadModels just once?

Reply:
You are right, i should move it in a separate bean config that will load models once upon startup and provide them in runtime whenever they are needed.